satsense/features/sift.py

```diff
@@ -5,33 +5,50 @@
 import numpy as np
 from sklearn.cluster import MiniBatchKMeans
 
+from ..generators import FullGenerator
 from ..image import Image
 from .feature import Feature
 
-SIFT = cv2.xfeatures2d.SIFT_create()
-
 
-def sift_cluster(images: Iterator[Image], n_clusters=32,
-                 sample_size=100000) -> MiniBatchKMeans:
+def sift_cluster(images: Iterator[Image],
+                 n_clusters=32,
+                 max_samples=100000,
+                 sample_window=(8192, 8192)) -> MiniBatchKMeans:
     """Create the clusters needed to compute the sift feature."""
-    descriptors = None
+    nfeatures = int(max_samples / len(images))
+    sift_object = cv2.xfeatures2d.SIFT_create(nfeatures)
+
+    descriptors = []
     for image in images:
-        array = image['gray_ubyte']
-        inverse_mask = (~array.mask).astype(np.uint8)
-        _, new_descriptors = SIFT.detectAndCompute(array, inverse_mask)
-        del _  # Free up memory
-
-        # Add descriptors if we already had some
-        if descriptors is None:
-            descriptors = new_descriptors
-        else:
-            descriptors = np.append(descriptors, new_descriptors, axis=0)
-
-    if descriptors.shape[0] > sample_size:
-        # Limit the number of descriptors to sample_size
-        # by randomly selecting some rows
-        descriptors = descriptors[np.random.choice(
-            descriptors.shape[0], sample_size, replace=False), :]
+        chunk = np.minimum(image.shape, sample_window)
+
+        generator = FullGenerator(image, chunk)
+        generator.load_image('gray_ubyte', (chunk, ))
+
+        max_features_per_window = int(nfeatures / np.prod(generator.shape))
+
+        img_descriptors = None
+        for img in generator:
+            inverse_mask = (~img.mask).astype(np.uint8)
+            _, new_descriptors = sift_object.detectAndCompute(
+                img, inverse_mask)
+            del _  # Free up memory
+
+            if new_descriptors.shape[0] > max_features_per_window:
+                new_descriptors = new_descriptors[np.random.choice(
+                    new_descriptors.shape[0],
+                    max_features_per_window,
+                    replace=False), :]
+
+            # Add descriptors if we already had some
+            if img_descriptors is None:
```

Comment:
Get rid of `img_descriptors` and directly append to `descriptors`. `np.append` is very inefficient because apparently it creates a copy of the array every time you use it.